examples/mnist/mnist-common.cpp

```diff
@@ -443,81 +527,141 @@ mnist_eval_result mnist_model_eval(const mnist_model & model, const float * imag
+    ggml_build_backward_expand(model.ctx_compute, gf, gb_grad, /*accumulate =*/ true, false);
 
-    struct ggml_opt_context opt_ctx;
-    struct ggml_opt_params  opt_pars = ggml_opt_default_params(GGML_OPT_TYPE_ADAM);
-    opt_pars.print_forward_graph = false;
-    opt_pars.print_backward_graph = false;
-    opt_pars.n_threads = nthreads;
-    opt_pars.adam.n_iter = 1; // per call of ggml_opt_resume_g
-    ggml_opt_init(model.ctx_compute, &opt_ctx, opt_pars, 0);
+    struct ggml_cgraph * gb_opt = ggml_graph_dup(model.ctx_compute, gf); // Backward pass, gradients + optimizer.
+    ggml_build_opt_adamw(model.ctx_compute, gf, gb_opt, 1e-3f, 0.9f, 0.999f, 1e-8f, 0.0f);
 
-    for (int epoch = 0; epoch < 20; ++epoch) {
-        fprintf(stderr, "%s: epoch %d start...", __func__, epoch);
+    model.buf_compute = ggml_backend_alloc_ctx_tensors(model.ctx_compute, model.backend);
+    ggml_graph_reset(gb_opt); // Set gradients to zero, reset optimizer.
+
+    const int iex_split = ((int)((1.0f - val_split)*nex) / model.nbatch_logical) * model.nbatch_logical;
+
+    for (int epoch = 0; epoch < nepoch; ++epoch) {
+        fprintf(stderr, "%s: epoch %02d start...", __func__, epoch);
         const int64_t t_start_us = ggml_time_us();
-        mnist_eval_result result;
-        for (int iex0 = 0; iex0 < nex; iex0 += model.nbatch) {
-            memcpy(model.images->data, images + iex0*MNIST_NINPUT,   ggml_nbytes(model.images));
-            memcpy(model.labels->data, labels + iex0*MNIST_NCLASSES, ggml_nbytes(model.labels));
 
-            enum ggml_opt_result opt_result = ggml_opt_resume_g(model.ctx_compute, &opt_ctx, model.loss, gf, gb, NULL, NULL);
-            GGML_ASSERT(opt_result == GGML_OPT_RESULT_OK || opt_result == GGML_OPT_RESULT_DID_NOT_CONVERGE);
+        float loss;
+        std::vector<float> logits(model.nbatch_physical*MNIST_NCLASSES);
+        int iex0 = 0;
+
+        mnist_eval_result result_train;
+        for (; iex0 < iex_split; iex0 += model.nbatch_physical) {
+            ggml_backend_tensor_set(model.images, images + iex0*MNIST_NINPUT,   0, ggml_nbytes(model.images));
+            ggml_backend_tensor_set(model.labels, labels + iex0*MNIST_NCLASSES, 0, ggml_nbytes(model.labels));
+
+            ggml_backend_graph_compute(model.backend, gf); // Always compute forward pass.
+
+            // With a period of nbatch_logical/nbatch_physical iterations:
+            if ((iex0 + model.nbatch_physical) % model.nbatch_logical != 0) {
+                // For the first nbatch_logical/nbatch_physical - 1 iterations, only calculate gradients and accumulate them:
+                ggml_backend_graph_compute(model.backend, gb_grad);
+            } else {
+                // For the last iteration, calculate gradients and also apply the optimizer:
+                ggml_backend_graph_compute(model.backend, gb_opt);
+                ggml_graph_reset(gb_grad); // Set gradients to zero, do not reset optimizer.
+            }
```

Comment:
This would never call `ggml_backend_graph_compute(model.backend, gb_grad);` if `nbatch_logical == nbatch_physical`

Reply:
It's not necessary. `gb_opt` contains all nodes in `gb_grad` so the gradients will still be calculated prior to the optimization step; I added a comment to clarify.